File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @param {number} n
     * @return {ListNode}
     */
    removeNthFromEnd(head: ListNode | null, n: number): ListNode {
        const dummy = new ListNode(0, head);
        let first: ListNode | null = dummy;
        let second: ListNode | null = dummy;

        for (let i = 0; i <= n; i++) {
            first = first!.next;
        }

        while (first !== null) {
            first = first.next;
            second = second!.next;
        }

        second!.next = second!.next!.next;

        return dummy.next as ListNode;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-3.go
func reorderList(head *ListNode) {
    slow, fast := head, head.Next
    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
    }

    second := slow.Next
    slow.Next = nil
    var prev *ListNode
    for second != nil {
        temp := second.Next
        second.Next = prev
        prev = second
        second = temp
    }
    second = prev

    first := head
    for second != nil {
        temp1, temp2 := first.Next, second.Next
        first.Next = second
        second.Next = temp1
        first = temp1
        second = temp2
    }
}